core/src/main/scala/kafka/coordinator/DelayedHeartbeat.scala

```diff
@@ -29,6 +29,8 @@ private[coordinator] class DelayedHeartbeat(coordinator: GroupCoordinator,
                                             heartbeatDeadline: Long,
                                             sessionTimeout: Long)
   extends DelayedOperation(sessionTimeout) {
+
+  override def safeTryComplete(): Boolean = tryComplete()
```

Comment:
Could we add a comment why safeTryComplete() needs to be customized in this case? Ditto in DelayedJoin.